core/src/main/java/com/scalar/db/transaction/consensuscommit/Snapshot.java

```diff
@@ -178,7 +181,9 @@ private TableMetadata getTableMetadata(Scan scan) throws ExecutionException {
   }
 
   public Optional<List<Key>> get(Scan scan) {
-    if (isWriteSetOverlappedWith(scan)) {
+    if (!ScalarDbUtils.isRelational(scan) && isWriteSetOverlappedWith(scan)) {
+      // Verify overlaps only for non-relational scan since we do it later for relational scan
+      // (right before returning the results).
       throw new IllegalArgumentException("reading already written data is not allowed");
     }
```

Comment:
I think we can move this overlap check logic into the newly added `verify()` method for consistency and readability. This would execute this check logic after the actual read, but I don't think this is a problem. What do you think?

Reply:
Let me confirm a bit. Do you mean that we can do the overlap check after the actual scan for both cases, regardless of whether it is relational or not? If yes, the answer is yes; we can do it without losing correctness. My only concern was unnecessary additional latency, but I think the readability would have more benefits here if we assume skilled users usually know it is prohibited and can avoid it.

Reply:
>  Do you mean that we can do the overlap check after the actual scan for both cases, regardless of whether it is relational or not?

Yes, correct.

> My only concern was unnecessary additional latency, but I think the readability would have more benefits here if we assume skilled users usually know it is prohibited and can avoid it.

The `IllegalArgumentException` issue is basically a bug, so maybe I don't think we need to worry about the additional latency. So in this case, I think we can prioritize the readability.

Reply:
OK, I will revise it. Thanks for the clarification and advice.

Reply:
Fixed in 31d2d2b.